README.md
Luxtronik2
==========
[![npm version](https://badge.fury.io/js/luxtronik2.svg)](https://badge.fury.io/js/luxtronik2)
[![Build Status](https://travis-ci.org/coolchip/luxtronik2.svg?branch=master)](https://travis-ci.org/coolchip/luxtronik2)
[![Dependency Status](https://david-dm.org/coolchip/luxtronik2.svg)](https://david-dm.org/coolchip/luxtronik2)
[![Codacy Badge](https://api.codacy.com/project/badge/Grade/d690e37c0ea94e8f99f1b1e36cd06687)](https://www.codacy.com/app/coolchip/luxtronik2?utm_source=github.com&amp;utm_medium=referral&amp;utm_content=coolchip/luxtronik2&amp;utm_campaign=Badge_Grade)
[![Code Climate](https://codeclimate.com/github/coolchip/luxtronik2/badges/gpa.svg)](https://codeclimate.com/github/coolchip/luxtronik2)
[![npm](https://img.shields.io/npm/l/express.svg)](https://www.npmjs.com/package/luxtronik2)

### Luxtronik2 reads and controls heat pumps based on the Luxtronik 2.0 contol unit with **Node.js**.

This work based on the fantastic [FHEM module 'LUXTRONIK2'](https://wiki.fhem.de/wiki/Luxtronik_2.0), the very usefull [openHAB binding 'Novelan Luxtronic heat pump'](
https://github.com/openhab/openhab1-addons/wiki/Novelan-Luxtronic-heat-pump-binding), the extensively [cbrandlehner/homebridge-luxtronik2](https://github.com/cbrandlehner/homebridge-luxtronik2) and a little bit research of my own. Hope you will like it.

#### Supports the following heat pumps
* Alpha Innotec
* Siemens Novelan (WPR NET)
* Roth (ThermoAura(r), ThermoTerra)
* Elco
* Buderus (Logamatic HMC20, HMC20 Z)
* Nibe (AP-AW10)
* Wolf Heiztechnik (BWL/BWS) 

How to use
----------
Connect your unit via lan and configure the ip parameters at your unit. The port number of your unit is 8888 by default.
Clone this code via git or simply via npm. You will get a package named luxtronik. You can require this at your code to
read and write to your heat pump.

Install
-------
Simply
```bash
npm install luxtronik2
```

Examples
--------
```javascript
var luxtronik = require('luxtronik2');

const hostIp = '127.0.0.1';  // <- Enter your Luxtronik IP here
var pump = new luxtronik.createConnection(hostIp, 8888);

// read all readable data
pump.read(function (err, data) {
    if (err) {
        return console.log(err);
    }
    console.log(data);
    console.log(data.values.errors);
});

// set heating target temperature to 0 °C
pump.write('heating_target_temperature', 0);

// set warm water target temperature to 60 °C and use callback
pump.write('warmwater_target_temperature', 60, function (err, res) {
    if (err) {
        return console.log(err);
    }
    console.log(res);
});

// set heating operation mode to 'Auto'
pump.write('heating_operation_mode', 0);

// set warm water operation mode to 'Auto'
pump.write('warmwater_operation_mode', 0);

// set heating target temperature and use callback
pump.write('heating_target_temperature', 0, function (err, res) {
    if (err) {
        return console.log(err);
    }
    console.log(res);
});

```

Migrating to version 1.0.0
--------------------------
The API changed between version 0.1.2 and version 1.0.0. [See migrating guide](MIGRATING.md) for information on how to migrate your application to the new API.


CHANGELOG.md
Upcoming
========
  * optimize file/folder structure
  * queue read/write jobs (fix for job collisions)

v1.0.0 / 25 Jul 2017 [see migrating guide](MIGRATING.md) 
========================================================
  * create object with luxtronik.createConnection() instead of "new"
  * return errors in callback instead of winston.log them
  * use error-first standard for handling callback parameters
  * correct spelling of "temperature" at write function
  * add early returns for cleaner functions
  * add installation how-to and rewrite example code
  * update dev-dependecies

v0.1.2 / 13 Apr 2017
====================
  * enable strict mode in all js files (**thanks to marcus**)

v0.1.1 / 13 Apr 2017
====================
  * insert roadmap
  * use of luxtronik always creates a new object - even without "new"
  * moved some code, renamed some functions and rewrite some passages to make the code more readable
  * add eslint to project and fix some eslint problems

v0.1.0 / 12 Apr 2017
====================
  * added missing error event handlers for client connection (**thanks to marcus**)
  * fixed handling, when heatpump is busy (**thanks to marcus**)
  * return plain text for error code and outage codes (**thanks to bakito**)

v0.0.2 / 11 Apr 2017
====================
  * add readRaw() function to get the raw data from the pump
  * code correctness

v0.0.1 / 06 Apr 2017
====================
  * first official release

test/luxtronikTest.js
'use strict';

const fs = require('fs');
const path = require('path');

const assert = require('assertthat');
const mitm = require('mitm')();

const luxtronik = require('../luxtronik');

let receiveDataBuffer = [];

mitm.on('connection', function (socket) {
    socket.on('data', function (data) {
        const recvSym = data.readInt32BE(0);
        if (recvSym === 0) {
            const command = receiveDataBuffer.pop();
            const dataFile = path.join(__dirname, 'data', command.toString());
            fs.readFile(dataFile, function (err, data) {
                if (!err) {
                    socket.write(data);
                }
            });
            receiveDataBuffer = [];
        } else {
            receiveDataBuffer.push(recvSym);
        }
    });
});

suite('Mocha tests', () => {
    suite('Luxtronik', () => {
        test('is an object.', done => {
            assert.that(luxtronik).is.ofType('object');
            done();
        });

        test('createConnection() function returns an object.', done => {
            assert.that(luxtronik.createConnection()).is.ofType('object');
            done();
        });

        test('connection has function read.', done => {
            assert.that(luxtronik.createConnection().read).is.ofType('function');
            done();
        });

        test('connection has function write.', done => {
            assert.that(luxtronik.createConnection().write).is.ofType('function');
            done();
        });
    });

    suite('Luxtronik pump', () => {
        const pump = luxtronik.createConnection('127.0.0.1', 8888);
        test('read returns data.', done => {
            pump.read(function (err, data) {
                assert.that(err).is.null();
                assert.that(data).is.ofType('object');
                assert.that(data.values).is.ofType('object');
                assert.that(data.parameters).is.ofType('object');
                done();
            });
        });
    });
});

utils.js
'use strict';

const humanizeduration = require('humanize-duration');
const huminizeoptions = {
    language: 'de',
    conjunction: ' und ',
    serialComma: false
};

const types = require('./types');

function createFirmwareString(buf) {
    let firmware = '';
    for (const key in buf) {
        if ({}.hasOwnProperty.call(buf, key)) {
            firmware += (buf[key] === 0) ? '' : String.fromCharCode(buf[key]);
        }
    }
    return firmware;
}

function int2ipAddress(value) {
    const part1 = value & 255;
    const part2 = ((value >> 8) & 255);
    const part3 = ((value >> 16) & 255);
    const part4 = ((value >> 24) & 255);
    return part4 + '.' + part3 + '.' + part2 + '.' + part1;
}

function createStateString(values) {
    let stateStr = '';
    const state1 = values[117];
    const state2 = values[118];
    const duration = values[120];

    // Text aus Define
    if (types.stateMessages.hasOwnProperty(state1)) {
        stateStr = types.stateMessages[state1];
        if (state2 === 0 || state2 === 2) {
            stateStr += ' seit ';
        } else if (state2 === 1) {
            stateStr += ' in ';
        }

        // Sonderbehandlung bei WP-Fehlern - Zeitstempel des zuletzt aufgetretenen Fehlers nehmen
        if (state2 === 2) {
            stateStr += new Date(values[95] * 1000).toString();
        } else {
            stateStr += humanizeduration(duration * 1000, huminizeoptions);
        }
    } else {
        stateStr = 'Unknown [' + state1 + ']';
    }
    return stateStr;
}

function createExtendedStateString(values) {
    let stateStr = '';
    const defrostValve = values[37];
    const heatSourceMotor = values[43];
    const compressor1 = values[44];
    const state3 = values[119];
    const ahpStufe = values[121];
    const ahpTemp = values[122] / 10;

    if (types.extendetStateMessages.hasOwnProperty(state3)) {
        stateStr = types.extendetStateMessages[state3];
        if (state3 === 6) {
            // Estrich Programm
            stateStr += ' Stufe ' + ahpStufe + ' - ' + ahpTemp + ' °C';
        } else if (state3 === 7) {
            // Abtauen
            if (defrostValve === 1) {
                stateStr += 'Abtauen (Kreisumkehr)';
            } else if (compressor1 === 0 && heatSourceMotor === 1) {
                stateStr += 'Luftabtauen';
            } else {
                stateStr += 'Abtauen';
            }
        }
    } else {
        stateStr = 'Unknown [' + state3 + ']';
    }
    return stateStr;
}

function createOperationStateString(state) {
    let stateStr = '';
    if (types.hpMode.hasOwnProperty(state)) {
        stateStr = types.hpMode[state];
    } else {
        stateStr = 'Unknown [' + state + ']';
    }
    return stateStr;
}

function createHotWaterStateString(values) {
    let stateStr = '';
    const hotWaterBoilerValve = values[38];
    const opStateHotWater = values[124];
    if (opStateHotWater === 0) {
        stateStr = 'Sperrzeit';
    } else if (opStateHotWater === 1 && hotWaterBoilerValve === 1) {
        stateStr = 'Aufheizen';
    } else if (opStateHotWater === 1 && hotWaterBoilerValve === 0) {
        stateStr = 'Temp. OK';
    } else if (opStateHotWater === 3) {
        stateStr = 'Aus';
    } else {
        stateStr = 'Unknown [' + opStateHotWater + '/' + hotWaterBoilerValve + ']';
    }
    return stateStr;
}

function createCode(time, code, codeTypes) {
    return {
        code,
        date: new Date(time * 1000),
        message: codeTypes.hasOwnProperty(code) ? codeTypes[code] : codeTypes[-1]
    };
}

function createCodeList(timeArray, codeArray, codeTypes) {
    const logArray = [];
    for (let i = 0; i < timeArray.length; i++) {
        logArray.push(createCode(timeArray[i], codeArray[i], codeTypes));
    }
    return logArray;
}

function createOutageCodeList(timeArray, codeArray) {
    return createCodeList(timeArray, codeArray, types.outageCodes);
}

function createErrorCodeList(timeArray, codeArray) {
    return createCodeList(timeArray, codeArray, types.errorCodes);
}

function toInt32ArrayReadBE(buffer) {
    const i32a = new Int32Array(buffer.length / 4);
    for (let i = 0; i < i32a.length; i++) {
        i32a[i] = buffer.readInt32BE(i * 4);
    }
    return i32a;
}

function createHeatPumptTypeString(value) {
    return types.hpTypes.hasOwnProperty(value) ? types.hpTypes[value] : types.hpTypes[-1];
}

function value2LuxtronikSetValue(realValue) {
    // Allow only integer temperature. Add factor x10.
    return parseInt(realValue * 10, 10);
}

function isValidOperationMode(value) {
    return types.hpMode.hasOwnProperty(value.toString());
}

function limitRange(value, min, max) {
    if (value < min) {
        value = min;
    }
    if (value > max) {
        value = max;
    }
    return value;
}

module.exports = {
    createFirmwareString,
    int2ipAddress,
    createStateString,
    createExtendedStateString,
    createOperationStateString,
    createHotWaterStateString,
    createOutageCodeList,
    createErrorCodeList,
    toInt32ArrayReadBE,
    createHeatPumptTypeString,
    value2LuxtronikSetValue,
    isValidOperationMode,
    limitRange
};


luxtronik.js
/**
 * @module luxtronik2
 * @copyright Sebastian B. <coolchip@gmx.de>
 */

'use strict';

const net = require('net');
const winston = require('winston');
winston.level = 'warning';

const utils = require('./utils');
const types = require('./types');

function Luxtronik(host, port) {
    if (!(this instanceof Luxtronik)) {
        return new Luxtronik(host, port);
    }

    if (typeof port === 'undefined') {
        this._port = 8888;
    } else {
        this._port = port;
    }
    this._host = host;
    this.receivy = {};
}

function processValues(heatpumpValues, heatpumpVisibility) {
    return {
        'temperature_supply': heatpumpValues[10] / 10, // #15
        'temperature_return': heatpumpValues[11] / 10, // #16
        'temperature_target_return': heatpumpValues[12] / 10, // #17
        'temperature_extern_return': (heatpumpVisibility[24] === 1) ? heatpumpValues[13] / 10 : 'no', // #18
        'temperature_hot_gas': heatpumpValues[14] / 10, // #26
        'temperature_outside': heatpumpValues[15] / 10, // #12
        'temperature_outside_avg': heatpumpValues[16] / 10, // #13
        'temperature_hot_water': heatpumpValues[17] / 10, // #14
        'temperature_hot_water_target': heatpumpValues[18] / 10, // #25
        'temperature_heat_source_in': heatpumpValues[19] / 10, // #23
        'temperature_heat_source_out': heatpumpValues[20] / 10, // #24
        'temperature_mixer1_flow': (heatpumpVisibility[31] === 1) ? heatpumpValues[21] / 10 : 'no', // #55
        'temperature_mixer1_target': (heatpumpVisibility[32] === 1) ? heatpumpValues[22] / 10 : 'no', // #56
        'temperaturw_RFV': (heatpumpVisibility[33] === 1) ? heatpumpValues[23] / 10 : 'no',
        'temperature_mixer2_flow': (heatpumpVisibility[34] === 1) ? heatpumpValues[24] / 10 : 'no', // #57
        'temperature_mixer2_target': (heatpumpVisibility[35] === 1) ? heatpumpValues[25] / 10 : 'no', // #48
        'temperature_solar_collector': (heatpumpVisibility[36] === 1) ? heatpumpValues[26] / 10 : 'no', // #50
        'temperature_solar_storage': (heatpumpVisibility[37] === 1) ? heatpumpValues[27] / 10 : 'no', // #51
        'temperature_external_source': (heatpumpVisibility[38] === 1) ? heatpumpValues[28] / 10 : 'no',

        'ASDin': heatpumpValues[29],
        'BWTin': heatpumpValues[30],
        'EVUin': heatpumpValues[31],
        'HDin': heatpumpValues[32],
        'MOTin': heatpumpValues[33],
        'NDin': heatpumpValues[34],
        'PEXin': heatpumpValues[35],
        'SWTin': heatpumpValues[36],

        'AVout': heatpumpValues[37],
        'BUPout': heatpumpValues[38],
        'HUPout': heatpumpValues[39],
        'MA1out': heatpumpValues[40],
        'MZ1out': heatpumpValues[41],
        'VENout': heatpumpValues[42],
        'VBOout': heatpumpValues[43],
        'VD1out': heatpumpValues[44],
        'VD2out': heatpumpValues[45],
        'ZIPout': heatpumpValues[46],
        'ZUPout': heatpumpValues[47],
        'ZW1out': heatpumpValues[48],
        'ZW2SSTout': heatpumpValues[49],
        'ZW3SSTout': heatpumpValues[50],
        'FP2out': heatpumpValues[51],
        'SLPout': heatpumpValues[52],
        'SUPout': heatpumpValues[53],
        'MZ2out': heatpumpValues[54],
        'MA2out': heatpumpValues[55],

        'defrostValve': (heatpumpVisibility[47] === 1) ? heatpumpValues[37] : 'no', // #67
        'hotWaterBoilerValve': heatpumpValues[38], // #9
        'heatingSystemCircPump': (heatpumpValues[39] === 1) ? 'on' : 'off', // #27

        'heatSourceMotor': (heatpumpVisibility[54] === 1) ? heatpumpValues[43] : 'no', // #64
        'compressor1': heatpumpValues[44],

        'hotWaterCircPumpExtern': (heatpumpVisibility[57] === 1) ? heatpumpValues[46] : 'no', // #28

        'hours_compressor1': Math.round(heatpumpValues[56] / 3600),
        'starts_compressor1': heatpumpValues[57],
        'hours_compressor2': Math.round(heatpumpValues[58] / 3600),
        'starts_compressor2': heatpumpValues[59],
        'hours_2nd_heat_source1': (heatpumpVisibility[84] === 1) ? Math.round(heatpumpValues[60] / 3600) : 'no', // #32
        'hours_2nd_heat_source2': (heatpumpVisibility[85] === 1) ? Math.round(heatpumpValues[61] / 3600) : 'no', // #38
        'hours_2nd_heat_source3': (heatpumpVisibility[86] === 1) ? Math.round(heatpumpValues[62] / 3600) : 'no', // #39
        'hours_heatpump': (heatpumpVisibility[87] === 1) ? Math.round(heatpumpValues[63] / 3600) : 'no', // #33
        'hours_heating': (heatpumpVisibility[195] === 1) ? Math.round(heatpumpValues[64] / 3600) : 'no', // #34
        'hours_warmwater': (heatpumpVisibility[196] === 1) ? Math.round(heatpumpValues[65] / 3600) : 'no', // #35
        'hours_cooling': (heatpumpVisibility[197] === 1) ? Math.round(heatpumpValues[66] / 3600) : 'no',

        'Time_WPein_akt': heatpumpValues[67],
        'Time_ZWE1_akt': heatpumpValues[68],
        'Time_ZWE2_akt': heatpumpValues[69],
        'Timer_EinschVerz': heatpumpValues[70],
        'Time_SSPAUS_akt': heatpumpValues[71],
        'Time_SSPEIN_akt': heatpumpValues[72],
        'Time_VDStd_akt': heatpumpValues[73],
        'Time_HRM_akt': heatpumpValues[74],
        'Time_HRW_akt': heatpumpValues[75],
        'Time_LGS_akt': heatpumpValues[76],
        'Time_SBW_akt': heatpumpValues[77],

        'typeHeatpump': utils.createHeatPumptTypeString(heatpumpValues[78]), // #31
        'bivalentLevel': heatpumpValues[79], // #43

        'WP_BZ_akt': heatpumpValues[80],

        'firmware': utils.createFirmwareString(heatpumpValues.slice(81, 91)), // #20

        'AdresseIP_akt': utils.int2ipAddress(heatpumpValues[91]),
        'SubNetMask_akt': utils.int2ipAddress(heatpumpValues[92]),
        'Add_Broadcast': utils.int2ipAddress(heatpumpValues[93]),
        'Add_StdGateway': utils.int2ipAddress(heatpumpValues[94]),

        'errors': utils.createErrorCodeList(heatpumpValues.slice(95, 100), heatpumpValues.slice(100, 105)), // #42 Time of first error

        'error_count': heatpumpValues[105],

        'switch_off': utils.createOutageCodeList(heatpumpValues.slice(111, 116), heatpumpValues.slice(106, 111)),

        'Comfort_exists': heatpumpValues[116],

        'heatpump_state1': heatpumpValues[117],
        'heatpump_state2': heatpumpValues[118], // #40
        'heatpump_state3': heatpumpValues[119],
        'heatpump_duration': heatpumpValues[120], // #41
        'heatpump_state_string': utils.createStateString(heatpumpValues),
        'heatpump_extendet_state_string': utils.createExtendedStateString(heatpumpValues),

        'ahp_Stufe': heatpumpValues[121],
        'ahp_Temp': heatpumpValues[122],
        'ahp_Zeit': heatpumpValues[123],

        'opStateHotWater': heatpumpValues[124], // #8
        'opStateHotWaterString': utils.createHotWaterStateString(heatpumpValues),
        'opStateHeating': heatpumpValues[125], // #46
        'opStateMixer1': heatpumpValues[126],
        'opStateMixer2': heatpumpValues[127],
        'Einst_Kurzprogramm': heatpumpValues[128],
        'StatusSlave_1': heatpumpValues[129],
        'StatusSlave_2': heatpumpValues[130],
        'StatusSlave_3': heatpumpValues[131],
        'StatusSlave_4': heatpumpValues[132],
        'StatusSlave_5': heatpumpValues[133],

        'rawDeviceTimeCalc': new Date(heatpumpValues[134] * 1000).toString(), // #22

        'opStateMixer3': heatpumpValues[135],
        'temperature_mixer3_target': (heatpumpVisibility[211] === 1) ? heatpumpValues[136] / 10 : 'no', // #60
        'temperature_mixer3_flow': (heatpumpVisibility[210] === 1) ? heatpumpValues[137] / 10 : 'no', // #59

        'MZ3out': heatpumpValues[138],
        'MA3out': heatpumpValues[139],
        'FP3out': heatpumpValues[140],

        'heatSourceDefrostTimer': (heatpumpVisibility[219] === 1) ? heatpumpValues[141] : 'no', // #66

        'Temperatur_RFV2': heatpumpValues[142] / 10,
        'Temperatur_RFV3': heatpumpValues[143] / 10,
        'SH_SW': heatpumpValues[144],
        'Zaehler_BetrZeitSW': Math.round(heatpumpValues[145] / 3600),
        'FreigabKuehl': heatpumpValues[146],
        'AnalogIn': heatpumpValues[147],
        'SonderZeichen': heatpumpValues[148],
        'SH_ZIP': heatpumpValues[149],
        'WebsrvProgrammWerteBeobarten': heatpumpValues[150],

        'thermalenergy_heating': (heatpumpVisibility[0] === 1) ? heatpumpValues[151] / 10 : 'no', // #36
        'thermalenergy_warmwater': (heatpumpVisibility[1] === 1) ? heatpumpValues[152] / 10 : 'no', // #37
        'thermalenergy_pool': (heatpumpVisibility[2] === 1) ? heatpumpValues[153] / 10 : 'no', // #62
        'thermalenergy_total': heatpumpValues[154] / 10,

        'analogOut1': heatpumpValues[156],
        'analogOut2': heatpumpValues[157],
        'Time_Heissgas': heatpumpValues[158],
        'Temp_Lueftung_Zuluft': heatpumpValues[159] / 10,
        'Temp_Lueftung_Abluft': heatpumpValues[160] / 10,

        'hours_solar': (heatpumpVisibility[248] === 1) ? Math.round(heatpumpValues[161] / 3600) : 'no', // #52
        'analogOut3': heatpumpValues[162],
        'analogOut4': (heatpumpVisibility[267] === 1) ? heatpumpValues[163] : 'no', // #73 - Voltage heating system circulation pump

        'Out_VZU': heatpumpValues[164],
        'Out_VAB': heatpumpValues[165],
        'Out_VSK': heatpumpValues[166],
        'Out_FRH': heatpumpValues[167],
        'AnalogIn2': heatpumpValues[168],
        'AnalogIn3': heatpumpValues[169],
        'SAXin': heatpumpValues[170],
        'SPLin': heatpumpValues[171],
        'Compact_exists': heatpumpValues[172],
        'Durchfluss_WQ': heatpumpValues[173],
        'LIN_exists': heatpumpValues[174],
        'LIN_TUE': heatpumpValues[175],
        'LIN_TUE1': heatpumpValues[176],
        'LIN_VDH': heatpumpValues[177],
        'LIN_UH': heatpumpValues[178],
        'LIN_UH_Soll': heatpumpValues[179],
        'LIN_HD': heatpumpValues[180],
        'LIN_ND': heatpumpValues[181],
        'LIN_VDH_out': heatpumpValues[182]
    };
}

function processParameters(heatpumpParameters, heatpumpVisibility) {
    return {
        'heating_temperature': heatpumpParameters[1] / 10, // #54 - returnTemperatureSetBack
        'warmwater_temperature': heatpumpParameters[2] / 10,
        'heating_operation_mode': heatpumpParameters[3], // #10
        'warmwater_operation_mode': heatpumpParameters[4], // #7

        'heating_operation_mode_string': utils.createOperationStateString(heatpumpParameters[3]),
        'warmwater_operation_mode_string': utils.createOperationStateString(heatpumpParameters[4]),

        'heating_curve_end_point': (heatpumpVisibility[207] === 1) ? heatpumpParameters[11] / 10 : 'no', // #69
        'heating_curve_parallel_offset': (heatpumpVisibility[207] === 1) ? heatpumpParameters[12] / 10 : 'no', // #70
        'deltaHeatingReduction': heatpumpParameters[13] / 10, // #47

        'heatSourcedefrostAirThreshold': (heatpumpVisibility[97] === 1) ? heatpumpParameters[44] / 10 : 'no', // #71

        'hotWaterTemperatureHysterese': heatpumpParameters[74] / 10, // #49

        'returnTempHyst': (heatpumpVisibility[93] === 1) ? heatpumpParameters[88] / 10 : 'no', // #68

        'heatSourcedefrostAirEnd': (heatpumpVisibility[105] === 1) ? heatpumpParameters[98] / 10 : 'no', // #72

        'temperature_hot_water_target': heatpumpParameters[105] / 10,

        'cooling_operation_mode': heatpumpParameters[108],

        'cooling_release_temperature': heatpumpParameters[110] / 10,
        'thresholdTemperatureSetBack': heatpumpParameters[111] / 10, // #48

        'cooling_inlet_temp': heatpumpParameters[132] / 10,

        'hotWaterCircPumpDeaerate': (heatpumpVisibility[167] === 1) ? heatpumpParameters[684] : 'no', // #61

        'heatingLimit': heatpumpParameters[699], // #11
        'thresholdHeatingLimit': heatpumpParameters[700] / 10, // #21

        'cooling_start_after_hours': heatpumpParameters[850],
        'cooling_stop_after_hours': heatpumpParameters[851],

        'typeSerial': heatpumpParameters[874].toString().substr(0, 4) + '/' + heatpumpParameters[874].toString().substr(4) + '-' + heatpumpParameters[875].toString(16).toUpperCase(),

        'returnTemperatureTargetMin': heatpumpParameters[979] / 10 // #63

        // "possible_temperature_hot_water_limit1": heatpumpParameters[47] / 10,
        // "possible_temperature_hot_water_limit2": heatpumpParameters[84] / 10,
        // "possible_temperature_hot_water_limit3": heatpumpParameters[973] / 10,
    };
}

Luxtronik.prototype._processData = function () {
    const heatpumpParameters = utils.toInt32ArrayReadBE(this.receivy['3003'].payload);
    const heatpumpValues = utils.toInt32ArrayReadBE(this.receivy['3004'].payload);
    const heatpumpVisibility = this.receivy['3005'].payload;

    if (typeof heatpumpParameters === 'undefined' ||
        typeof heatpumpValues === 'undefined' ||
        typeof heatpumpVisibility === 'undefined') {
        return this.receivy.callback(new Error('Unexpected Data'));
    }
    if (this.receivy.rawdata) {
        return this.receivy.callback(null, {
            values: '[' + heatpumpValues + ']',
            parameters: '[' + heatpumpParameters + ']'
        });
    }

    const values = processValues(heatpumpValues, heatpumpVisibility);
    const parameters = processParameters(heatpumpParameters, heatpumpVisibility);
    const additional = {
        'reading_calculated_time_ms': this.receivy.readingEndTime - this.receivy.readingStartTime
    };

    // flow rate
    values.flowRate = (heatpumpParameters[870] !== 0) ? heatpumpValues[155] : 'no'; // #19

    // skips inconsistent flow rates (known problem of the used flow measurement devices)
    if (values.flowRate !== 'no' && values.heatingSystemCircPump === 'on') {
        if (values.flowRate === 0) {
            values.flowRate = 'inconsistent';
        }
    }

    if (parameters.hotWaterCircPumpDeaerate !== 'no') {
        parameters.hotWaterCircPumpDeaerate = parameters.hotWaterCircPumpDeaerate ? 'on' : 'off';
    }

    // Consider also heating limit
    let heatingStateString = '';
    if (parameters.heating_operation_mode === 0 && parameters.heatingLimit === 1 &&
        values.temperature_outside_avg >= parameters.thresholdHeatingLimit &&
        (values.temperature_target_return === parameters.returnTemperatureTargetMin ||
            values.temperature_target_return === 20 && values.temperature_outside < 10)
    ) {
        if (values.temperature_outside >= 10) {
            heatingStateString = 'Heizgrenze (Soll ' + parameters.returnTemperatureTargetMin + ' °C)';
        } else {
            heatingStateString = 'Frostschutz (Soll 20 °C)';
        }
    } else {
        if (types.heatingState.hasOwnProperty(values.opStateHeating)) {
            heatingStateString = types.heatingState[values.opStateHeating];
        } else {
            heatingStateString = 'unbekannt (' + values.opStateHeating + ')';
        }

        // Consider heating reduction limit
        if (values.opStateHeating === 0) {
            if (parameters.thresholdTemperatureSetBack <= values.temperature_outside) {
                heatingStateString += ' ' + parameters.deltaHeatingReduction + ' °C';
            } else {
                heatingStateString = 'Normal da < ' + parameters.thresholdTemperatureSetBack + ' °C';
            }
        }
    }
    values.opStateHeatingString = heatingStateString;

    return this.receivy.callback(null, {
        values,
        parameters,
        additional
    });
};

function sendData(client, data) {
    if (typeof client !== 'undefined' && client !== null) {
        data.forEach(function (element) {
            const buffer = Buffer.allocUnsafe(4);
            buffer.writeInt32BE(element);
            client.write(buffer);
        });
    }
}

Luxtronik.prototype._nextJob = function () {
    if (this.receivy.jobs.length > 0) {
        this.receivy.activeCommand = 0;
        sendData(this.client, [this.receivy.jobs.shift(), 0]);
    } else {
        this.client.end();
        this.client = null;
        this.receivy.readingEndTime = Date.now();
        process.nextTick(this._processData.bind(this));
    }
};

Luxtronik.prototype._startRead = function (rawdata, callback) {
    this.receivy = {
        jobs: [3003, 3004, 3005],
        activeCommand: 0,
        readingStartTime: Date.now(),
        rawdata,
        callback
    };

    this.client = net.createConnection({
        host: this._host,
        port: this._port
    }, function () {
        winston.log('debug', 'Connected');
        process.nextTick(this._nextJob.bind(this));
    }.bind(this));

    this.client.on('error', function (error) {
        winston.log('error', error);
        this.client.end();
        this.client = null;
        process.nextTick(
            function () {
                this.receivy.callback(error);
            }.bind(this)
        );
    }.bind(this));

    this.client.on('data', function (data) {
        if (this.receivy.activeCommand === 0) {
            const commandEcho = data.readInt32BE(0);
            let firstReadableDataAddress = 0;

            if (commandEcho === 3004) {
                const status = data.readInt32BE(4);
                if (status > 0) {
                    winston.log('error', 'Parameter on target changed, restart parameter reading after 5 seconds');
                    this.client.end();
                    this.client = null;
                    return process.nextTick(
                        function () {
                            this.receivy.callback(new Error('heatpump busy'));
                        }.bind(this)
                    );
                } else {
                    firstReadableDataAddress = 12;
                }
            } else {
                firstReadableDataAddress = 8;
            }
            const paramCount = data.readInt32BE(firstReadableDataAddress - 4);
            let dataCount = 0;
            if (commandEcho === 3005) {
                // 8 Bit values
                dataCount = paramCount;
            } else {
                // 32 Bit values
                dataCount = paramCount * 4;
            }
            const payload = data.slice(firstReadableDataAddress, data.length);

            this.receivy.activeCommand = commandEcho;
            this.receivy[commandEcho] = {
                remaining: dataCount - payload.length,
                payload
            };
        } else {
            this.receivy[this.receivy.activeCommand] = {
                remaining: this.receivy[this.receivy.activeCommand].remaining - data.length,
                payload: Buffer.concat([this.receivy[this.receivy.activeCommand].payload, data])
            };
        }

        if (this.receivy[this.receivy.activeCommand].remaining <= 0) {
            winston.log('debug', this.receivy.activeCommand + ' completed');
            process.nextTick(this._nextJob.bind(this));
        }
    }.bind(this));

    this.client.on('close', function () {
        winston.log('debug', 'Connection closed');
    });
};

Luxtronik.prototype._startWrite = function (setParameter, setValue, callback) {
    this.client = net.createConnection({
        host: this._host,
        port: this._port
    }, function () {
        winston.log('debug', 'Connected');
        const command = 3002;
        sendData(this.client, [command, setParameter, setValue]);
    }.bind(this));

    this.client.on('error', function (error) {
        winston.log('error', error);
        process.nextTick(
            function () {
                callback(error);
            }
        );
        this.client.end();
        this.client = null;
    }.bind(this));

    this.client.on('data', function (data) {
        const commandEcho = data.readInt32BE(0);
        let next;
        if (commandEcho !== 3002) {
            const error = 'Host did not confirm parameter setting';
            winston.log('error', error);
            next = function () {
                callback(error);
            };
        } else {
            const setParameterEcho = data.readInt32BE(4);
            winston.log('debug', setParameterEcho + ' - ok');
            next = function () {
                callback(null, {
                    msg: 'write ok'
                });
            };
        }
        process.nextTick(next);
        this.client.end();
        this.client = null;
    }.bind(this));
};

Luxtronik.prototype._handleWriteCommand = function (parameterName, realValue, callback) {
    const writeParameters = Object.freeze({
        'heating_target_temperature': {
            setParameter: 1,
            setValue: utils.value2LuxtronikSetValue(utils.limitRange(realValue, -10, 10))
        },
        'warmwater_target_temperature': {
            setParameter: 2,
            setValue: utils.value2LuxtronikSetValue(utils.limitRange(realValue, 30, 65))
        },
        'heating_operation_mode': {
            setParameter: utils.isValidOperationMode(realValue) ? 3 : undefined,
            setValue: realValue
        },
        'warmwater_operation_mode': {
            setParameter: utils.isValidOperationMode(realValue) ? 4 : undefined,
            setValue: realValue
        },
        'cooling_operation_mode': {
            setParameter: 108,
            setValue: realValue
        },
        'cooling_release_temp': {
            setParameter: 110,
            setValue: utils.value2LuxtronikSetValue(realValue)
        },
        'cooling_inlet_temp': {
            setParameter: 132,
            setValue: utils.value2LuxtronikSetValue(realValue)
        },
        'cooling_start': {
            setParameter: 850,
            setValue: realValue
        },
        'cooling_stop': {
            setParameter: 851,
            setValue: realValue
        },
        'wrongName': {
            //setParameter: undefined,
        }
    });

    const set = writeParameters.hasOwnProperty(parameterName) ? writeParameters[parameterName] : writeParameters.wrongName;
    if (typeof set.setParameter !== 'undefined') {
        const setParameter = set.setParameter;
        const setValue = set.setValue;
        winston.log('debug', 'Set parameter ' + parameterName + '(' + setParameter + ') = ' + realValue + '(' + setValue + ')');
        this._startWrite(setParameter, setValue, callback);
    } else {
        const error = 'Wrong data';
        winston.log('error', error);
        process.nextTick(
            function () {
                callback(error);
            }
        );
    }
};

Luxtronik.prototype.read = function (rawdata, callback) {
    if (rawdata instanceof Function) {
        callback = rawdata;
        rawdata = false;
    }
    this._startRead(rawdata, callback);
};

Luxtronik.prototype.readRaw = function (callback) {
    this._startRead(true, callback);
};

Luxtronik.prototype.write = function (parameterName, realValue, callback) {
    if (typeof callback === 'undefined') {
        callback = function () {};
    }
    this._handleWriteCommand(parameterName, realValue, callback);
};

const createConnection = function (host, port) {
    return new Luxtronik(host, port);
};

module.exports.createConnection = createConnection;